sqlglot/optimizer/simplify.py

```diff
@@ -547,7 +553,34 @@ def simplify_literals(expression, root=True):
 NULL_OK = (exp.NullSafeEQ, exp.NullSafeNEQ, exp.PropertyEQ)
 
 
+def _simplify_integer_cast(expr: exp.Expression) -> exp.Expression:
+    if isinstance(expr, exp.Cast) and isinstance(expr.this, exp.Cast):
+        this = _simplify_integer_cast(expr.this)
```

Comment:
what happens if it's like

cast(cast(x as int) as bool) ?

Reply:
Good question; In an intermediate implementation I had created a closure called `is_integer_cast` that checked if the cast type is of `INTEGER_TYPE`, in order to make sure that the cast above wouldn't even recurse. However, I decided that it's probably superfluous; To illustrate, let's unroll the calls:

- On the first invocation of `simplify_integer_cast`, we see a double cast, so we recursively call it again on the first branch
- On the second invocation, we see `CAST(1 AS INT)` which is a viable optimization in our case, thus we return / bubble up the value `Literal(1)`. 
- Coming back up on the first invocation, we now have `this = Literal(1)` making the expression to `CAST(1 AS BOOL)`. That ends up being rejected as it's incorrect; Note that the recursive calls don't change the arguments in place, thus when the rejection happens on the top call, `return expr` makes sure to return the unoptimized version of the "top" expression as-is in order to repair the comparison. 

Is that enough to cover these cases, or should I add explicit checks to make sure that the cast type is of `INTEGER_TYPES` (which would also prune a function call as well)? Also, I thought I had actually added a test showcasing exactly that e.g. with nested casts to `TEXT`, but I must have removed it, will add a few now.

Reply:
Good question, and even though I had tested this locally and included a test case for it, I somehow can't find it so I'll make sure to include a few "faulty" ones on the next iteration. My thinking was that this case is handled as is, so let's unroll the calls to see it:

- On the first invocation of `_simplify_integer_cast`, we see a nested cast, thus we recurse
- On the second invocation of `_simplify_integer_cast`, we now see `CAST(1 as INT)`. Here, the cast is removed because the conditions are met, and thus we return / bubble-up the value `Literal(1)`
- Coming back on the first invocation, we now have `this = Literal(1)` so semantically the top-level expression is now equivalent to `CAST(1 AS BOOLEAN)`. This, of course, does not meet the criteria so it's "rejected" and we return the whole `expr` as is. Note that the recursion calls don't alter the arguments in place, so by preserving the nodes we've "repaired" the whole expression with no cost.

This is not to be confused with the canonicalize rule of `remove_redundant_cast` that on later optimization passes would actually remove `CAST(1 AS INT)`. If the nested casts where e.g. `CAST(CAST(1 AS BIGINT) AS BOOL)`, then we'd observe that both simplify and canonicalize would leave it as is.